src/dotnet/ToolPackage/ToolPackageInstance.cs

```diff
@@ -166,5 +172,28 @@ private static LockFileItem FindItemInTargetLibrary(LockFileTargetLibrary librar
                 ?.ToolsAssemblies
                 ?.SingleOrDefault(t => LockFileMatcher.MatchesFile(t, targetRelativeFilePath));
         }
+
+        public IEnumerable<string> Warnings
```

Comment:
This property does non-trivial work.  I think it should either use a `Lazy<IEnumerable<string>>` under the covers or this should really be a method called `GetWarnings`.

Reply:
Also, we already deserialized the configuraiton when we accessed the commands.  Perhaps this and the `Commands` property should share an underlying `Lazy<ToolConfiguration>`?

Reply:
I tried to scope the access of this property as explicit as possible. Since `warnings` is not a concern for rest of the component (unless we can predict the future).

Reply:
```c#
private Lazy<ToolConfiguration> _toolConfiguration = new Lazy<ToolConfiguration>(GetToolConfiguration);

public IEnumerable<string> Warnings => _toolConfiguration.Value.Warnings;

private ToolConfiguration GetToolConfiguration()
{
    try
    {
        var lockFile = new LockFileFormat().Read(PackageDirectory.WithFile(AssetsFileName).Value);
        var library = FindLibraryInLockFile(lockFile);
        return DeserializeToolConfiguration(ToolSettingsFileName, library);
    }
    catch (Exception ex) when (ex is UnauthorizedAccessException || ex is IOException)
    {
        throw new ToolConfigurationException(
            string.Format(
                CommonLocalizableStrings.FailedToRetrieveToolConfiguration,
                ex.Message),
            ex);
    }
}
```

And then use `_toolConfiguration.Value` in `GetCommands`.

Reply:
So i will go GetWarnings. it is pretty much only ever accessed once. I would like Lazy if the syntax is less noisy

Reply:
so who ever access ToolConfiguration or Warnings first will trigger GetToolConfiguration()?

My concern is access Warnings will trigger other unrelated field get populated. No strong feeling, if you think it is good, let's go this route. It can reduce a lot of code

Reply:
Accessing `Warnings` first won't trigger anything to be populated in `Commands`, just save it the time it would have spent on deserializing the tool configuration a second time, is all.

Reply:
it needs to deserialize the whole thing anyway. I guess i will just go as you suggested